### PR TITLE
core: add DEFAULT_APP_ENTRY to select main() function

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR/RFC adds a mechanism to change the default entry point to the user main function.
To change main entry point simply define DEFAULT_APP_ENTRY to a different function (by default, will call main() ).

### Why?
I found this use cases:
* For JavaScript apps (#7796), the main() function is (somehow) just an event_loop() call and could be a common js_main() inside sys/js folder.
* For Arduino support, we could have an arduino_main() that handles setup() and loop() functions
* If at some point we add an App framework (provide custom setup, loop, handlers, whatever), we won't need to touch core files.

I also added a mechanism to print the name of the function, which I'm still not 100% sure if it's the way to go (message should still be there?, etc).

All feedback will be really appreciated!

### CODE REVIEW

File: core/kernel_init.c [hunk 1]
```diff
@@ -54,8 +63,7 @@ static void *main_trampoline(void *arg)
     stat->laststart = 0;
 #endif
 
-    LOG_INFO("main(): This is RIOT! (Version: " RIOT_VERSION ")\n");
-
+    LOG_INFO(DEFAULT_APP_ENTRY_NAME "(): This is RIOT! (Version: " RIOT_VERSION ")\n");
     main();
```
Comment (gebart) [inline core/kernel_init.c]: Missed this

Comment (jia200x) [reply core/kernel_init.c]: fixed

**Comment by kaspar030** [comment]
I don't quite agree with the use cases. Using this define, one other function can be selected (instead of main).

This could simple be solved by the corresponding modules just defining "main()", instead of an application.

**Comment by tcschmidt** [comment]
@jia200x do you agree with the objections of @kaspar030 or shall we ask for another review?

**Comment by jia200x** [comment]
@tcschmidt I agree. Let's close this PR and continue in favour of #9105

**Pull request closed without merging**